src/System.Collections.Sequences/System/Collections/Sequences/Position.cs

```diff
@@ -10,39 +10,59 @@ namespace System.Collections.Sequences
     public struct Position : IEquatable<Position>
     {
         long _index;
-        object _obj;
+        object _item;
 
         public static explicit operator int(Position position) => (int)position._index;
-        public static implicit operator long(Position position) => position._index;
-        public T As<T>() => _obj == null || IsInfinity ? default : (T)_obj;
-        public bool IsInfinity => this == Infinity;
 
-        public static readonly Position Infinity = new Position(new object(), long.MaxValue);
+        public static implicit operator long(Position position) => position._index;
 
-        public static Position operator +(Position position, long value) => new Position(position._obj, position._index + value);
+        public long GetIndex() => _index;
 
-        public static Position operator -(Position position, long value) => new Position(position._obj, position._index - value);
+        public T GetItem<T>() => _item == null || IsEnd ? default : (T)_item;
 
-        public static bool operator ==(Position left, Position right) => left._index == right._index && left._obj == right._obj;
-        public static bool operator !=(Position left, Position right) => left._index != right._index || left._obj != right._obj;
+        public (T item, long index) Get<T>() => (GetItem<T>(), this);
 
-        public void Set(long value) => _index = value;
+        public void SetIndex(long index) => _index = index;
 
-        public void Set(int value) => _index = value;
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void SetItem<T>(T item) where T:class
```

Comment:
nit: spacing

Also, why do we need to constraint T to a class here/elsewhere?

Reply:
I will fix the spacing. 
The constraint is here so that people don't accidentally box and so the null check inside the method can be efficient